File: Duplicata/Palindromo.java
import java.util.Scanner;

public class Palindromo {

    public static boolean palindromo(String string) {
        int tam, iguais = 0;

        tam = string.length();

        // Remove o '\n' caso haja na entrada
        if (string.charAt(tam - 1) == '\n') {
            tam--; // Apenas ajusta o tamanho da string, sem precisar usar substring
        }

        // Cria uma nova string sem espaços
        String copiaSemEspacos = "";
        for (int i = 0; i < tam; i++) {
            if (string.charAt(i) != ' ') {
                copiaSemEspacos += string.charAt(i); // Concatena os caracteres
            }
        }

        int tam2 = copiaSemEspacos.length();
        String stringInvertida = "";

        // Cria a string invertida
        for (int i = tam2 - 1; i >= 0; i--) {
            stringInvertida += copiaSemEspacos.charAt(i);
        }

        // Compara as duas strings ignorando maiúsculas e minúsculas
        for (int i = 0; i < tam2; i++) {
            char c1 = copiaSemEspacos.charAt(i);
            char c2 = stringInvertida.charAt(i);

            // Compara de forma case-insensitive (transforma manualmente para maiúsculas)
            if ((c1 >= 'a' && c1 <= 'z') && (c2 >= 'a' && c2 <= 'z')) {
                c1 -= 'a' - 'A'; // Converte c1 para maiúsculo
            } else if ((c1 >= 'A' && c1 <= 'Z') && (c2 >= 'a' && c2 <= 'z')) {
                c2 -= 'a' - 'A'; // Converte c2 para maiúsculo
            }

            if (c1 == c2) {
                iguais++;
            }
        }

        return iguais == tam2; // Retorna true se todas as letras forem iguais
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        String entrada;

        do {
            entrada = scanner.nextLine();

            // Remove o \n manualmente
            if (entrada.length() > 0 && entrada.charAt(entrada.length() - 1) == '\n') {
                entrada = entrada.substring(0, entrada.length() - 1); // Ajuste para manter a regra
            }

            if (palindromo(entrada)) {
                System.out.println("SIM");
            } else {
                System.out.println("NAO");
            }

        } while (!entrada.equals("FIM")); // Verifica se a entrada é "FIM"

        scanner.close();
    }
}

File: Duplicata/palindromo.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <ctype.h>

bool palindromo(char string[])
{
    int j = 0, tam, iguais = 0;

    tam = strlen(string);

    if (string[tam - 1] == '\n') // remove o '\n' do final da string e coloca um '\0'
    {
        string[tam - 1] = '\0';
    }

    char copia_sem_espacos[tam];

    for (int i = 0; i < (tam - 1); i++) // loop que faz a copia_sem_espacos da string de entrada para uma outra sem os espacos
    {
        if (string[i] != ' ')
        {
            copia_sem_espacos[j] = string[i];
            j++;
        }
    }
    copia_sem_espacos[j] = '\0'; // finaliza a string com um \0 ao final

    int tam2 = strlen(copia_sem_espacos);
    j = tam2 - 1;
    char string_invertida[tam2 + 1];

    for (int i = 0; i < tam2; i++) // loop para copiar a string sem os espacos par outra string de forma que inverta a posicao dos caracteres
    {
        string_invertida[i] = copia_sem_espacos[j];
        j--;
    }
    string_invertida[tam2] = '\0';

    for (int i = 0; i < tam2; i++) // loop para verificar se a string se mantem igual mesmo apos inverte-la
    {
        if (toupper(string_invertida[i]) == toupper(copia_sem_espacos[i])) // torna todos os caracteres da string maiusculos, assim nao havera diferenca no momento de verificacao
        {
            iguais++;
        }
    }

    if (iguais == tam2) // com uma variavel auxiliar que conta quantos caracteres sao iguais, verifiquei se as duas strings possuem a mesma quantia de caracteres
    {
        return true; // se sim e um palindromo e retorna true
    }
    else
    {
        return false; // se nao, nao e um palindromo, e retorna false
    }
}

int main()
{
    char string[100];

    do
    {
        fgets(string, sizeof(string), stdin);

        if (strcmp(string, "FIM") != 0) // Verifica a condição de parada
        {
            if (palindromo(string))
            {
                printf("SIM\n");
            }
            else
            {
                printf("NAO\n");
            }
        }
    } while (strcmp(string, "FIM") != 0);
}